### ISSUE TITLE
Add greenlet

### ISSUE THREAD

**Issue by rth** [pyodide/pyodide#2664]
Recently @antocuni mentioned that it could be worth exploring the idea of [green threads](https://en.wikipedia.org/wiki/Green_threads) in relationship with https://github.com/pyodide/pyodide/issues/1503 and https://discuss.python.org/t/wasm-python-synchronous-io-working-group/15509.

The first step would likely be to add the [greenlet](https://greenlet.readthedocs.io/en/latest/). The build setup  is in https://github.com/pyodide/pyodide/commit/7bda00302363f3ea444a2e37a04d782bd95c6a34 however this currently fails with, 
```
src/greenlet/greenlet.c:374:6: error: "greenlet needs to be ported to this platform, 
or taught how to detect your compiler properly."
```
so it looks one should add a WASM platform implementation [here](https://github.com/python-greenlet/greenlet/tree/master/src/greenlet/platform). The implementations for other platforms are not very long, but quite low level.

I don't know much about this topic, but I found it interesting that greenlet is often used in combination with gevent which is an API on top of the [libev](http://software.schmorp.de/pkg/libev.html) (inspired by libevent) or [libuv](http://libuv.org/) event loops. While the last two are event loops used e.g. in Chromium and Node.js respectively.

**Comment by hoodmane** [comment]
Related: [wasm stack switching proposal](https://github.com/WebAssembly/stack-switching/blob/main/proposals/stack-switching/Overview.md)

**Comment by antocuni** [comment]
Some more thoughts about this. My initial idea is that once we have greenlets, it should be easy/possible to simulate synchronous calls from Python. E.g., I can imagine having an implementation of `requests.get` which under the hoods uses a greenlet to stop its execution and yield back the control to the browser. So, it would be sync from the perspective of the python user and async from the browser point of view.

In order to support greenlets, we need some sort of C level stack switching.
I'm new to the webassembly/emscripten world and I'm still trying to wrap my head around it, but from what I understand there are at least three different approaches:
- [asyncify](https://emscripten.org/docs/porting/asyncify.html): if I understand correctly, when using this emscripten rewrites the WASM in order to support stack-switching, and it adds some overhead. Moreover, I fear that it would need rewrite the *whole* CPython in order to work, and probably all the extensions would need to be asyncified as well. But again, I might have misunderstood
- [emscripten WASM stack switching](https://github.com/emscripten-core/emscripten/pull/16779): this looks more promising but it's also pretty new, we need to investigate how well it works
- the built-in wasm stack switching proposal mentioned by @hoodmane : this would be the best, but if I understand correctly it's just a proposal for now.

Once we have greenlets, we need a way to integrate them with the browser's event loop; these two projects look relevant:
- https://github.com/miguelgrinberg/greenletio
- https://github.com/gfmio/asyncio-gevent

**Comment by rth** [comment]
Thanks a lot for looking into this @antocuni ! This direction sounds very promising, and if this could be made to work it would indeed be a significantly lighter solution compared to web-workers. 

> [asyncify](https://emscripten.org/docs/porting/asyncify.html):

There was some exploration of this in https://github.com/pyodide/pyodide/issues/1503#issuecomment-944898189 it looks the CPython works to some extent. But overall yes, a significant overhead on performance (though we would likely need to measure it more accurately) doesn't sound like a very good compromise to get this feature.

> emscripten WASM stack switching: this looks more promising

I agree, it's also promising that Emscripten devs work on this and so if we find issues with the current implementation it's likely that they would be receptive to improving it.  Hood started work to update to emscripten 3.1.13 in https://github.com/pyodide/pyodide/pull/2668 which would be necessary to try it out. 

> if I understand correctly it's just a proposal for now.

That's what I understood as well. The discussions in https://github.com/WebAssembly/stack-switching/issues/2 and https://github.com/WebAssembly/meetings/tree/main/stack/2021 are quite interesting, and I wonder if the Emscripten implementation is somewhat a result of those discussions.

**Comment by hoodmane** [comment]
> emscripten WASM stack switching: this looks more promising

That's really exciting! I think it shouldn't be *too* hard to get emscripten up to 3.1.13. All of our cleanup has really made updating Emscripten a lot more manageable. Our biggest blocker was the firefox bug but that seems to be cleared now because people update firefox.

**Comment by hoodmane** [comment]
Unfortunately, it seems that it is not possible to implement the `slp_switch` function for wasm even with wasm stack switching. The difficulty is that greenlet doesn't just switch call stacks, it actually duplicates the call stack so that part of it returns twice. This is similar to the way that `fork` returns twice: the call stack from the parent process is duplicated into the child so both experience returning from fork into the same call frame.

Most likely the only way to get greenlet to work is to make significant modifications.

We definitely cannot do this in wasm with the JS Promise Integration proposal: we are allowed to suspend a call stack but we *cannot* duplicate it and return from the same function twice. Maybe the full wasm-stack-switching proposal could help when it is completed by doing some complicated thing with a two-stage loop and a suspend instruction but it is unclear even in that case. I think we would still need to do a code transform to add this two-stage loop up the call stack.

I think it could be possible to get greenlet to work by doing very careful surgery on the compiled webassembly module. The key point is that only a few functions are used on both the original and the duplicated call stack. 
1. Each function that executes both in the original greenlet and in the new greenlet needs to be modified so that all local variables are spilled to the argument stack since we can access the duplicated argument stack in the spawned greenlet but we can only access the true call stack in the parent greenlet.
2. Once we've insured that all locals are spilled to the arg stack, we can duplicate the segment of the function from the call to the stack switcher downwards. In the `slp_switch` function, we call into an appropriate truncated bottom half function after preparing the arg stack appropriately so that it sees the correct values of the spilled locals.

Since there are multiple code paths that land in `slp_switch`, we would need some way of storing the function pointer to the entry point too so that `slp_switch` knows which truncated code path to invoke. As part of the transformation, they could be modified to insert this information onto the argument stack.

It would be an interesting project to investigate this, but it seems quite hard. Maybe @jamadden can provide advice.

**Comment by antocuni** [comment]
Uhm, I admit that I don't fully understand *everything* here and I am probably missing some pieces.
But I wonder whether the approach used by PyPy can be useful here.
In PyPy, greenlets are implemented in pure Python on top of a lower-level primitive called `continulet` (written in RPython).
Here is a [brief description](https://foss.heptapod.net/pypy/pypy/-/blob/branch/default/pypy/module/_continuation/moduledef.py) of its semantics. If we can implement `continulet` on top of WASM stack switching, we can probably re-use PyPy's own implementation of `greenlet`.

The pure python implementation of greenlet based on contnulet is in [greenlet.py](https://foss.heptapod.net/pypy/pypy/-/blob/branch/default/lib_pypy/greenlet.py)

The [WASM stack switching proposal](https://github.com/WebAssembly/stack-switching) explicitly says that it wants to support continuations, which seems to be closely related to PyPy's continulet: but I admit I don't really know.

Reusing the code of `continulet` directly is probably not possible, because it ends up calling `slp_switch` again.
`continulet` are implemented on top of an even lower-level primitive called [stacklet](https://foss.heptapod.net/pypy/pypy/-/blob/branch/default/rpython/translator/c/src/stacklet/stacklet.h) (written in C).
Stacklets seem to use code which was borrowed from the original greelnet, and in particular [here it calls `slp_switch](
https://foss.heptapod.net/pypy/pypy/-/blob/branch/default/rpython/translator/c/src/stacklet/stacklet.c#L251-284).

So, to summarize:
1. if we can implement continulet in WASM, we can probably re-use PyPy's greenlet.py
2. I doubt that we can reuse PyPy's `stacklet` code, which means that we would need to reimplement a WASM-specific version of `continulet`
3. I don't really know whether what I said so far makes any sense 😅.

@arigo is the original author of `continulet`, so maybe he knows better than me?

**Comment by hoodmane** [comment]
To be clear, my goal is to get this working on top of JS Promise integration, not the full wasm stack switching proposal since the latter is (a) not going to be ready for quite a bit longer and (b) much harder to use in a backwards compatible way. It is theoretically possible to implement green thread support just using JS Promise integration, but the question is whether it lines up cleanly with existing abstractions.

Based on the concepts in greenlet I think it will be reasonable to make my branch #3210 arbitrarily reentrant (currently it is nested reentrant but it can't handle a pair of interlocking loops). I think from there it shouldn't be difficult to implement continulet.

**Comment by hoodmane** [comment]
Okay, copying from greenlet did indeed allow me to fix the two interlocking Python loops case. For instance, the following code seems to work:
```js
let f = pyodide.runPython(`
  from js import sleep
  def f(x):
    n = 0
    while True:
      n += 1
      print(x, n)
      sleep(x).syncify()
  f
`);
f.callSyncifying(250);
f.callSyncifying(150);
```

**Comment by hoodmane** [comment]
Well I made an implementation of continulet that passes almost the entire test suite (there are a few tests that were too tightly coupled to pypy to port reasonably). However, `permute` doesn't work yet. Anyways I think I've done most of the work to get greenlets working.

The tests work in v8 version 10.8.168.11 but it seems to be broken in v8 version 10.7.193.13, so in particular it won't work with node version 19 which uses v8 version 10.7.

**Comment by antocuni** [comment]
So if I understand correctly, pyodide's greenlets will be implemented on top of continulet as we do on PyPy?

**Comment by ryanking13** [comment]
Thanks for your work @hoodmane. This is going to be a huge breakthrough for Pyodide. I am very excited to use it :) I haven't looked into the stack-switching proposal specs deeply yet, but I'll do it when I have time so I can review your PRs.

>  my goal is to get this working on top of JS Promise integration, not the full wasm stack switching proposal since the latter is (a) not going to be ready for quite a bit longer and (b) much harder to use in a backwards compatible way.

I like the idea of implementing this on top of JS promise integration. One thing I want to discuss is, the stack switching JS promise integration seems like it is in a [prototype state](https://chromestatus.com/feature/5674874568704000), and it seems like it didn't get an agreement from other browser vendors yet. So wouldn't there be a chance that it is discarded later?

I mean, we already have an API that is only available in specific browsers, so I am +1 with having a new API that is only available in Chrome. The question I have is: is it guaranteed that the Chrome team is not going to drop that feature later in future releases?

**Comment by hoodmane** [comment]
> So if I understand correctly, pyodide's greenlets will be implemented on top of continulet as we do on PyPy?

This is the plan. @antocuni Can you provide any extra information about what the `permute` function is supposed to do? I don't really understand how it is supposed to behave from the test suite.

**Comment by ryanking13** [comment]
> Well I don't think we should merge this before it either:
> 
> 1. works in node with some experimental flag or
> 2. works in Chrome with no flag

That sounds reasonable to me. Thanks!

**Comment by antocuni** [comment]
> This is the plan. @antocuni Can you provide any extra information about what the `permute` function is supposed to do? I don't really understand how it is supposed to behave from the test suite.

I never really understood continulets to be honest, and I personally did very few work on them, so I fear I don't have an easy answer.
But if you want I'm happy to do some pair programming and try to (re)understand them together

**Comment by davidbrochart** [comment]
Just asking, it seems that greenlet is still not supported in Pyodide?

**Comment by agriyakhetarpal** [comment]
Hi @davidbrochart, yes, it isn't.

**Comment by syrusakbary** [comment]
Some great updates!

We have been able to enable greenlets fully on WebAssembly! We detailed a bit more how here:
https://wasmer.io/posts/greenlet-support-python-wasm

**Comment by salabim** [comment]
Does that mean that we will (soon?) have greenlet supported in pyodide? That would be great.

**Comment by hoodmane** [comment]
Well since the wasmer API looks similar to JSPI, we could probably reuse a lot of their work in principle. It's not clear that wasmer's greenlet port is open source though.
